File: groupctl-prototype/volpanctl.py
from typing import Literal
from events import (
    Event,
    EventBus,
    EventBusSubscriber,
    PageSelected,
    PaintSideButton,
    SideButtonClick,
    PanModeSelected,
    SideButton,
    VolModeSelected,
)


VolPanMode = Literal["vol"] | Literal["pan"]


VOL_COLOR: Literal[69] = 69
PAN_COLOR: Literal[69] = 69


class VolPanCtl(EventBusSubscriber):
    def __init__(self, bus: EventBus) -> None:
        self.bus = bus
        self.bus.subscribe(self)
        self.mode: VolPanMode = "vol"
        self.page_active: bool = True
        self._paint()

    def _paint(self) -> None:
        if not self.page_active:
            return
        self.bus.send(
            PaintSideButton(
                button=SideButton.volume,
                color=VOL_COLOR if self.mode == "vol" else 0,
            )
        )
        self.bus.send(
            PaintSideButton(
                button=SideButton.pan,
                color=VOL_COLOR if self.mode == "pan" else 0,
            )
        )
        
    def on(self, event: Event) -> None:
        match event:
            case SideButtonClick(button=SideButton.volume):
                self.mode = "vol"
                self.bus.send(VolModeSelected())
                self._paint()
            case SideButtonClick(button=SideButton.pan):
                self.mode = "pan"
                self.bus.send(PanModeSelected())
                self._paint()
            case PageSelected(n=n):
                self.page_active = n == 0
                self._paint()

File: GroupCtl/src/main/java/dev/tradcode/groupctl/mixmachine/VolPanCtl.java
package dev.tradcode.groupctl.mixmachine;

import dev.tradcode.groupctl.events.Event;
import dev.tradcode.groupctl.events.IEventBus;
import dev.tradcode.groupctl.events.IEventBusSubscriber;
import dev.tradcode.groupctl.events.PageSelected;
import dev.tradcode.groupctl.events.PaintSideButton;
import dev.tradcode.groupctl.events.PanModeSelected;
import dev.tradcode.groupctl.events.SideButton;
import dev.tradcode.groupctl.events.SideButtonClick;
import dev.tradcode.groupctl.events.VolModeSelected;

enum VolPanMode { VOL, PAN };

public class VolPanCtl implements IEventBusSubscriber {
    IEventBus bus;
    boolean pageActive = true;
    VolPanMode mode = VolPanMode.VOL;

    public VolPanCtl(IEventBus bus) {
        this.bus = bus;
        this.bus.subscribe(this);
        this.paint();
    }

    public void paint() {
        if (!this.pageActive)
            return;
        this.bus.send(
            new PaintSideButton(
                SideButton.VOLUME,
                this.mode == VolPanMode.VOL ? MixMachineColors.VOL_COLOR : 0
            ),
            new PaintSideButton(
                SideButton.PAN,
                this.mode == VolPanMode.PAN ? MixMachineColors.PAN_COLOR : 0
            )
        );
    }

    public void on(Event event) {
        switch (event) {
            case SideButtonClick(var btn) when btn == SideButton.VOLUME -> {
                this.mode = VolPanMode.VOL;
                this.bus.send(new VolModeSelected());
                this.paint();
            }
            case SideButtonClick(var btn) when btn == SideButton.PAN -> {
                this.mode = VolPanMode.PAN;
                this.bus.send(new PanModeSelected());
                this.paint();
            }
            case PageSelected(int n) -> {
                this.pageActive = n == 0;
                this.paint();
            }
            default -> { }
        }
    }
}
